libs/cli/src/main/java/org/elasticsearch/cli/internal/JsonPrintWriter.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cli.internal;
+
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.time.Clock;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.Locale;
+
+/**
+ * {@link PrintWriter} formatting lines as JSON using a subset of fields written by ECSJsonLayout.
+ *
+ * When using {@code println}, anything that looks like a JSON object is immediately passed through and written as is.
+ */
+public class JsonPrintWriter extends PrintWriter {
+    private static final DateTimeFormatter ISO_INSTANT_FORMATTER = new DateTimeFormatterBuilder().appendInstant(3).toFormatter(Locale.ROOT);
+    static final int BUFFER_INITIAL_SIZE = 512;
+    static final int BUFFER_TRIM_THRESHOLD = 2 * BUFFER_INITIAL_SIZE;
+
+    // Buffer to build JSON mimicking the ECSJsonLayout.
+    // NEVER acquire monitor lock on `lock` if holding `buffer`, always acquire `lock` first!
+    private final StringBuilder buffer = new StringBuilder(BUFFER_INITIAL_SIZE);
+    private final Clock clock;
+
+    public JsonPrintWriter(OutputStream out, boolean autoFlush) {
+        this(out, autoFlush, Clock.systemUTC());
+    }
+
+    @SuppressForbidden(reason = "Override PrintWriter to emit Json")
+    protected JsonPrintWriter(OutputStream out, boolean autoFlush, Clock clock) {
+        super(out, autoFlush);
+        this.clock = clock;
+    }
+
+    private void initJsonIfEmpty() {
+        assert Thread.holdsLock(buffer); // only allow if holding lock on buffer
+        if (buffer.isEmpty()) {
+            buffer.append("{\"@timestamp\":\"");
+            ISO_INSTANT_FORMATTER.formatTo(clock.instant(), buffer);
+            buffer.append("\", \"message\":\"");
```

Comment:
I think there are some other fields we should fill in. eg `log.level`, `log.logger`. Perhaps others, are any other fields required by ECS (eg `ecs.version)? Will we be able to find these messages associated with a node if we don't output `elasticsearch.node.name` and `elasticsearch.cluster.name`? cc @pgomulka

Reply:
we should be able to still find those logs as filebeat will add additional fields like `docker.container.name` or `ece.cluster` `ece.instance`. in ESS cloud is still parsing .log files instead of json files, so it proves it is doable :) , but raises a question if we should stream the stdout/err to both plaintext log and json (if this change applies to all distributions)

`elasticsearch.node.name`  is dynamically set because it can be generated if no `node.name` is provided. So not sure you can easily use it here without logging dependencies.

`elasticsearch.cluster.name` is based on a static setting based on `cluster.name` setting. Might be worth setting that field.


node.id, cluster.uuid are both dynamically generated and stored in thread context - again it would be hard to access them here.

I agree we should set log.level and log.logger.
but also `event.dataset` = `elasticsearch.server` (or elasticserach.stdout?)

service.name = ES_ECS
perhaps ecs.version = 1.12.0

Reply:
> but raises a question if we should stream the stdout/err to both plaintext log and json

Note that this is the CLI process. I don't think we should try writing to log files. We are only trying to ensure that when ES is emitting json to stdout/stderr, the CLI also emits json. This can happen in two ways: (1) if the cli logs something and (2) if the ES process emits non-json directly from the jdk (this would normally only happen early in startup, before ES has replaced the stdout/stderr streams.

Reply:
>  ESS cloud is still parsing .log files instead of json files

Is that true? Sounds like they discard stderr / stdout of the launcher then?

Reply:
> I think there are some other fields we should fill in. eg log.level, log.logger. Perhaps others, are any other fields required by ECS 

That's part of the reason why I initially considered depending on ecs-logging-core. As @pgomulka pointed out some field require additionally logic defined in the logging dependencies.

- `logger`: that one is simple to add ✔️ 
- `level`: are we good with a static level per PrintWriter (INFO for stdout, ERROR for stderr)? PrintWriter isn't aware of the Terminal verbosity level, but required (for instance, the JSON print writer is passed to the options parser). So that would require some ThreadLocal hack then. 

I'll give this a closer look tomorrow 👍

Reply:
These look good to start with, we can consider more in followups. As for ecs version, I'm not sure how much it matters, maybe @pgomulka has more thoughts there.

Reply:
@pgomulka Could you give this another look? Do you think the fields are sufficient to start with? Is there any way I can test / verify this?

Reply:
>> ESS cloud is still parsing .log files instead of json files
> Is that true? Sounds like they discard stderr / stdout of the launcher then?

you mean now or after this change is merged?
currently stdout in ESS (not serverless) goes to console. They are still able to consume those logs though, you can find them by filtering for `stderr`. Bear in mind that logging is customised and configuration is in cloud repo

in serverless the ES's docker logging config is used, all output goes to stdout (except for a gc logs). Streams can be separated by checking the event.dataset (slow logs, deprecation etc. although nothing except for server is currently needed)

Reply:
> @pgomulka Could you give this another look? Do you think the fields are sufficient to start with? Is there any way I can test / verify this?

try at minimum expanding the ESJsonLogsConfigIT or create a different test that would start ES server and consume the logs.  There is also a unit test JsonLoggerTests.java, might be good to add a test case there too.

we should expand the test cases in filebeat for the logs we produce. Try create some happy/error path logs.
see https://github.com/elastic/beats/tree/main/filebeat/module/elasticsearch/server/test
Create a test sample and reach out to the beats team- they will help you create a PR or they will do this themselves

once you have this PR merged to main, you can use the snapshot build to test in staging ESS and see if the logs are being correctly ingested into logging clusters
the same for serverless, we should confirm that those logs are correctly being ingested into logging clusters in QA (you can deploy a custom snapshot build there)
I am not sure we have e2e tests to help testing this.

it is possible to test in DEV, but it might be time consuming to set up (unless there were improvements).